src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -344,32 +354,52 @@ private enum ParsingState : byte
             {
                 Debug.Assert(_connection != null);
 
-                int drainedBytes = 0;
-                while (true)
+                CancellationTokenSource cts = null;
+                CancellationTokenRegistration ctr = default;
+                try
                 {
-                    drainedBytes += _connection.RemainingBuffer.Length;
+                    int drainedBytes = 0;
                     while (true)
                     {
-                        ReadOnlyMemory<byte> bytesRead = ReadChunkFromConnectionBuffer(int.MaxValue);
-                        if (bytesRead.Length == 0)
+                        drainedBytes += _connection.RemainingBuffer.Length;
+                        while (true)
                         {
-                            break;
+                            ReadOnlyMemory<byte> bytesRead = ReadChunkFromConnectionBuffer(int.MaxValue, ctr);
+                            if (bytesRead.Length == 0)
+                            {
+                                break;
+                            }
                         }
-                    }
 
-                    // When ReadChunkFromConnectionBuffer reads the final chunk, it will clear out _connection
-                    // and return the connection to the pool.
-                    if (_connection == null)
-                    {
-                        return true;
-                    }
+                        // When ReadChunkFromConnectionBuffer reads the final chunk, it will clear out _connection
+                        // and return the connection to the pool.
+                        if (_connection == null)
+                        {
+                            return true;
+                        }
 
-                    if (drainedBytes >= maxDrainBytes)
-                    {
-                        return false;
-                    }
+                        if (drainedBytes >= maxDrainBytes)
+                        {
+                            return false;
+                        }
+
+                        if (cts == null) // only create the drain timer if we have to go async
+                        {
+                            TimeSpan drainTime = _connection._pool.Settings._maxResponseDrainTime;
```

Comment:
It would be nice if we could have drain cancellation logic in one place and just pass a cancellationToken to DrainAsync.  That said, I think we can't do this because we want to defer creating the token until we need to read from the network.  Am I understanding properly?

I suppose we could split DrainAsync into DrainBuffered and DrainAsync that took a cancellationToken.  Not sure that's worth it, just food for thought...

Reply:
> That said, I think we can't do this because we want to defer creating the token until we need to read from the network. Am I understanding properly?

Right.